Sources/NIOHPACK/SimpleRingBuffer.swift

```diff
@@ -0,0 +1,609 @@
+            self._moveTail(to: _storage.capacity - r + w)
+        }
+    }
+    
+    @inlinable
+    mutating func _moveHead(to newIndex: Int) {
+        assert(newIndex >= 0 && newIndex <= self._storage.capacity)
+        self._ringHead = newIndex
+    }
+    
+    @inlinable
+    mutating func _moveHead(forwardBy amount: Int) {
+        assert(amount >= 0)
+        let readerIsBelowWriter = self._ringHead <= self._ringTail
+        var newIndex = self._ringHead + amount
+        
+        // reader index should not pass writer index
+        if readerIsBelowWriter {
+            assert(newIndex <= self._ringTail)
+        } else if newIndex >= self._storage.capacity {
+            // wrap it around
+            newIndex -= self._storage.capacity
+            // now test that it didn't pass writer
+            assert(newIndex <= self._ringTail)
+        }
+        
+        self._moveHead(to: newIndex)
+        self._readableBytes -= amount
+    }
+    
+    @inlinable
+    mutating func _moveTail(to newIndex: Int) {
+        assert(newIndex >= 0 && newIndex <= self._storage.capacity)
+        self._ringTail = newIndex
+    }
+    
+    @inlinable
+    mutating func _moveTail(forwardBy amount: Int) {
+        assert(amount >= 0)
+        let writerIsBelowReader = self._ringTail < self._ringHead
+        var newIndex = self._ringTail + amount
+        
+        // writer index should not pass readerIndex
+        if writerIsBelowReader {
+            assert(newIndex <= self._ringHead)
+        } else if newIndex >= self._storage.capacity {
+            // wrap it around
+            newIndex -= self._storage.capacity
+            // now test it
+            assert(newIndex <= self._ringHead)
+        }
+        
+        self._moveTail(to: newIndex)
+        self._readableBytes += amount
+    }
+    
+    @inlinable
+    mutating func moveHead(forwardBy amount: Int) {
+        self._moveHead(forwardBy: amount)
+    }
+    
+    @inlinable
+    mutating func moveTail(forwardBy amount: Int) {
+        self._moveTail(forwardBy: amount)
+    }
+    
+    @inlinable
+    mutating func unwrite(byteCount: Int) {
+        assert(byteCount >= 0)
+        if byteCount <= self._ringTail {
+            self._ringTail -= byteCount
+        } else {
+            self._ringTail = self._storage.capacity - byteCount + self._ringTail
+        }
+        self._readableBytes -= byteCount
+    }
+    
+    @inlinable
+    mutating func unread(byteCount: Int) {
+        assert(byteCount >= 0)
+        if byteCount <= self._ringHead {
+            self._ringHead -= byteCount
+        } else {
+            self._ringHead = self._storage.capacity - byteCount + self._ringHead
+        }
+        self._readableBytes += byteCount
+    }
+    
+    @inlinable
+    mutating func _set<S: ContiguousCollection>(bytes: S, at index: Int) throws -> Int where S.Element == UInt8 {
+        let totalCount = bytes.count
+        guard totalCount <= self.writableBytes else {
```

Comment:
In fact, this `guard` suggests we don't allow doing this at arbitrary points, as this `guard` assumes that we will be making this buffer larger, even though there's no actual operation that does so here.

Reply:
The `guard` it designed to *prevent* the buffer from growing, no? It's ensuring it isn't asked to actually write more than it can accommodate.

Reply:
Sorry, this comment makes more sense when read in the context of [this one](
https://github.com/apple/swift-nio-http2/pull/10/files/c5089386640dcee25c1a3533fa33e480e7f86c0e#r204414209). What I'm getting at is that the `guard` here assumes that the value of `writableBytes` will be decremented by the length of this buffer, but then never does it. This `guard` will fire if I try to write 10 bytes into a buffer that only has 5 bytes free, even if I try to write it at a point that is entirely unrelated to the writer index.

I basically think the two `_set` methods are confused, and we don't want them. We want `write` methods that definitely move the writer index, and so can meaningfully make guard statements based on that, rather than have arbitrary set methods. The arbitrary set methods on ByteBuffer exist to allow you to disregard the writer/reader indices, but doing so on this buffer type is silly, so we don't want to support it.